app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt

```diff
@@ -38,76 +70,175 @@ class FireDatabase : Database() {
+                        future.completeExceptionally(java.lang.Error("Impossible to set this username !"))
+                    }
+                }
+            }
+        }
+        return future
+    }
+
+    override fun setUsername(username: String): CompletableFuture<Boolean> {
+        val future = CompletableFuture<Boolean>()
+
+        val userId = getUserId()
+        //check if the username is available
+        isUsernameAvailable(username).thenAccept{isAvailable ->
+            if(isAvailable){
+                //add the link between the username and the userId
+                val usernameToUserId = HashMap<String, String>()
+                usernameToUserId.put(username, userId)
+                database.child(usernameToUserIdFileName).updateChildren(usernameToUserId as Map<String, Any>)
+                    .addOnSuccessListener {
+                        //add the users account to the database and the username to the user account
+                        val userAccount = HashMap<String, String>()
+                        userAccount.put("username", username)
+                        database.child("users").child(userId).updateChildren(userAccount as Map<String, Any>)
+                            .addOnSuccessListener { future.complete(true) }
+                            .addOnFailureListener {
+                                future.completeExceptionally(Exception("Impossible to create the user account."))
+                            }
+                    }
+                    .addOnFailureListener {
+                        future.completeExceptionally(java.lang.Error("Impossible to find the link between the username and the userId to the Database.") )
+                    }
+            }else{
+                future.completeExceptionally(java.lang.Error("The username is not available !"))
+            }
+        }
+        return future
     }
 
-    override fun setUsername(username: String) {
-        val userAdd = HashMap<String, String>()
-        userAdd.put(username, "empty")
-        database.child("users").updateChildren(userAdd as Map<String, Any>)
-            .addOnFailureListener { throw Exception("Impossible to add the username on the database") }
+    override fun initUserProfile(userModel: UserModel): CompletableFuture<Boolean> {
+        val userData = HashMap<String, Any>()
+        userData.put("email", userModel.getEmailAddress())
+        userData.put("firstname", userModel.getFirstname())
+        userData.put("surname", userModel.getSurname())
+        userData.put("dateOfBirth", userModel.getDateOfBirth().toEpochDay())
+        userData.put("distanceGoal", userModel.getDistanceGoal())
+        userData.put("activityTimeGoal", userModel.getActivityTime())
+        userData.put("nbOfPathsGoal", userModel.getNumberOfPathsGoal())
+
+        val future = CompletableFuture<Boolean>()
+        val userId = getUserId()
+
+        database.child(usersProfileFileName).child(userId).updateChildren(userData)
```

Comment:
`database.child(usersProfileFileName).child(userId)` is copied 4 times in this file, it should be refactored to its own function.